File: Min_cost_of_tickets.py
"""
TC:O(N) where N is the no. of elements in the list of days given 
SC:O(1)
"""
#This is a DP problem of exploring all paths
#minCost() calculates the min cost of tickets from all paths
#nd()() function calculates the no. of days after which the day pass gets invalid
class Solution:
    def mincostTickets(self, days: List[int], costs: List[int]) -> int:
        self.dp = [0] * len(days)
        self.days = days
        self.costs = costs
        return self.minCost(0)
        
    def nd(self, day_indx, duration):
        if duration == 1:
            return day_indx + 1
        
        indx = day_indx
        lastday = self.days[indx] + duration - 1
        
        while indx < len(self.days) and self.days[indx] <= lastday:
            indx += 1
            
        return indx
    
    def minCost(self, index):
        #base case
        if index == len(self.days):
            return 0
        
        if self.dp[index] != 0:
            return self.dp[index]
        
        #logic
        
        self.dp[index] = min(self.costs[0] + self.minCost(self.nd(index, 1)), self.costs[1] + self.minCost(self.nd(index, 7)), self.costs[2] + self.minCost(self.nd(index, 30)))
        
        return self.dp[index]
        

File: MinCostOfTickets.java
/*
Time complexity - O(N), where N is the last day in the days array.
Space complexity - O(N)

Approach - 
    Construct an array of size of the value of the last day in the days array.
    Add all days in the days array to set. 
    Iterate from one to farthest day.
    At each step 
                if the day is not in the input array, simply copy the cost of prev day.
                otherwise, calculate the following
                        a. single day pass - cost of prev day + cost of single day pass.
                        b. weekly pass - cost of prev week + cost of weekly pass.
                                         if prev week doesnt exist -> cost of weekly pass alone.
                        c. monthly pass - equivalent of weekly pass.
                        find the min of a,b,c.
                store it in result array.
    return cost for the last day / farthest day.
*/
class Solution {
    public int mincostTickets(int[] days, int[] costs) {
        return dp_bottom_up(days, costs);
        //return mincostTickets(days, costs, 0);
    }

    private int dp_bottom_up(int[] days, int[] costs) {
        Set<Integer> set = new HashSet<>();
        int lastDay = days[days.length-1], dp[] = new int[lastDay+1];

        for (int day : days) set.add(day);

        for (int i = 1; i <= lastDay; i++)
            if (!set.contains(i))
                dp[i] = dp[i-1];
            else {
                dp[i] = dp[i - 1] + costs[0];

                int j = (i >= 7) ? i - 7 : 0;
                dp[i] = Math.min(dp[i], dp[j] + costs[1]);

                j = (i >= 30) ? i - 30 : 0;
                dp[i] = Math.min(dp[i], dp[j] + costs[2]);
            }
        return dp[lastDay];
    }

    /*private int mincostTickets(int[] days, int[] costs, int cost) {
        Queue<Pair<int, int>> last7, last30;
        for (int d : days) {
            while (!last7.empty() && last7.peek().getKey() + 7 <= d) 
                last7.pop();
            while (!last30.empty() && last30.peek().getKey() + 30 <= d) 
                last30.pop();
            last7.push(new Pair(d, cost + costs[1]));
            last30.push(new Pair(d, cost + costs[2]));
            cost = min({cost + costs[0], last7.peek().getValue(), last30.peek().getValue() });
        }
        return cost;
    }*/
}
